assets/src/edit-story/components/panels/stylePreset/presets.js

```diff
@@ -0,0 +1,279 @@
+    }
+  };
+
+  useLayoutEffect(() => {
+    if (colorsRef.current && null !== activeColorIndex) {
+      colorsRef.current.querySelector('[tabindex="0"]').focus();
+    }
+  }, [activeColorIndex]);
+
+  useLayoutEffect(() => {
+    if (stylesRef.current && null !== activeStyleIndex) {
+      stylesRef.current.querySelector('[tabindex="0"]').focus();
+    }
+  }, [activeStyleIndex]);
+
+  useKeyDownEffect(
+    colorsRef,
+    { key: ['up', 'down', 'left', 'right'], shift: true },
+    ({ key }) => {
+      const maxIndex = isText ? textColors.length - 1 : fillColors.length - 1;
+      if (colorsRef.current) {
+        const diff = getIndexDiff(key, COLORS_PER_ROW);
+        const newIndex = Math.max(
+          0,
+          Math.min(maxIndex, (activeColorIndex ?? 0) + diff)
+        );
+        setActiveColorIndex(newIndex);
+      }
+    },
+    [activeColorIndex]
+  );
+
+  useKeyDownEffect(
+    stylesRef,
+    { key: ['up', 'down', 'left', 'right'], shift: true },
+    ({ key }) => {
+      if (stylesRef.current) {
+        const diff = getIndexDiff(key, STYLES_PER_ROW);
+        const newIndex = Math.max(
+          0,
+          Math.min(textStyles.length - 1, (activeStyleIndex ?? 0) + diff)
+        );
+        setActiveStyleIndex(newIndex);
+      }
+    },
+    [activeStyleIndex]
+  );
+
+  const getStylePresetText = (preset) => {
+    const isHighLight =
+      preset.backgroundTextMode === BACKGROUND_TEXT_MODE.HIGHLIGHT;
+    // @todo Confirm text content usage.
+    const text = textContent.substring(0, 8) + '...';
+    return (
+      <TextWrapper>
+        {isHighLight ? (
+          <HighLight background={preset.backgroundColor}>{text}</HighLight>
+        ) : (
+          text
+        )}
+      </TextWrapper>
+    );
+  };
+
+  const colorPresets = isText ? textColors : fillColors;
+  const hasColorPresets = colorPresets.length > 0;
+
+  const groupLabel = isText
+    ? __('Text colors', 'web-stories')
+    : __('Colors', 'web-stories');
+  return (
+    <PanelContent isPrimary padding={hasColorPresets ? null : '0'}>
+      {hasColorPresets && (
+        <>
+          <PresetGroupLabel>{groupLabel}</PresetGroupLabel>
+          <PresetGroup ref={colorsRef}>
+            {colorPresets.map((color, i) => (
+              <ButtonWrapper key={`color-${i}`}>
+                <Color
+                  tabIndex={
+                    activeColorIndex === i || (!activeColorIndex && i === 0)
+                      ? 0
+                      : -1
+                  }
+                  color={color}
+                  {...getEventHandlers(color)}
+                  aria-label={
+                    isEditMode
+                      ? __('Delete color preset', 'web-stories')
+                      : __('Apply color preset', 'web-stories')
+                  }
+                >
+                  {isEditMode && <Remove />}
+                </Color>
+              </ButtonWrapper>
+            ))}
+          </PresetGroup>
+        </>
```

Comment:
You have this repeated twice now - once for colors and once for styles. Can this be generalized to a common component - perhaps with a render prop as to how to render the individual items? A lot of the logic is duplicated, though there are still quite a few differences too of course.

Reply:
If you could change it to something like this, it would be nice:

```js
<PresetList
  list={colors}
  label={groupLabel}
  itemRenderer={(value, isSelected, onKeyDown}) => (
    <Color
      tabIndex={isSelected ? 0 : -1}
      color={value}
      onKeyDown={onKeyDown}
      onClick={() => somethingHandler(value)}
      aria-label={isEditMode ? __('Delete color preset', 'web-stories') : __('Apply color preset', 'web-stories')}
    >
       {isEditMode && <Remove />}
    </Color>
  )}
/>
```

Then all the keyboard handling logic, current index, etc, can be moved to this `<PresetList>` component and the panel component will be a lot cleaner.

Reply:
Hm, it does also require you to move some more logic out of the `StylePresetPanel` component and into `Presets` or `PresetList`. Maybe `Presets` should just be updated to work as the above, and the outer `StylePresetPanel` should then figure out whether to include one or two presets itself?

But 3 different components seem fine here too actually. I'd just make sure to move the logic for keeping and updating the current index to the innermost component where possible - it seems wasteful to duplicate that and put it in the outer one.

Reply:
Good point! Done in 76ad91dd